File: internal/trie/trie.go
package trie

type TrieNode struct {
	children    map[rune]*TrieNode
	isEndOfWord bool
}

type Trie struct {
	root *TrieNode
}

func NewTrie() *Trie {
	return &Trie{
		root: &TrieNode{children: make(map[rune]*TrieNode)},
	}
}

func (t *Trie) Populate(words []string) {
	if len(words) == 0 {
		return
	}
	for _, word := range words {
		t.Insert(word)
	}
}

func (t *Trie) Insert(word string) {
	if len(word) == 0 {
		return
	}
	curr := t.root
	for _, char := range word {
		if curr.children[char] == nil {
			curr.children[char] = &TrieNode{children: make(map[rune]*TrieNode)}
		}
		curr = curr.children[char]
	}
	curr.isEndOfWord = true
}

func (t *Trie) Delete(word string) bool {
	if t.root == nil || len(word) == 0 {
		return false
	}
	return t.deleteHelper(t.root, word, 0)
}

func (t *Trie) deleteHelper(node *TrieNode, word string, index int) bool {
	if index == len(word) {
		if !node.isEndOfWord {
			return false
		}
		node.isEndOfWord = false
		return len(node.children) == 0
	}

	char := rune(word[index])
	child, exists := node.children[char]
	if !exists {
		return false
	}

	shouldDeleteChild := t.deleteHelper(child, word, index+1)
	if shouldDeleteChild {
		delete(node.children, char)
		return len(node.children) == 0
	}
	return false
}

func (t *Trie) GetAllWords() []string {
	words := make([]string, 0)
	if len(t.root.children) == 0 {
		return []string{}
	}
	t.getAllWordsDfs(t.root, "", &words)
	return words
}

func (t *Trie) getAllWordsDfs(node *TrieNode, path string, words *[]string) {
	if node.isEndOfWord {
		*words = append(*words, path)
	}

	for char, child := range node.children {
		t.getAllWordsDfs(child, path+string(char), words)
	}
}

func (t *Trie) FuzzySearch(pattern string) []string {
	if len(pattern) == 0 {
		return []string{}
	}
	words := make([]string, 0)
	t.fuzzySearchDfs(t.root, pattern, "", 0, &words)
	return words
}
func (t *Trie) SearchPrefix(prefix string) []string {
	if len(prefix) == 0 {
		return []string{}
	}
	curr := t.root
	for _, char := range prefix {
		if curr.children[char] == nil {
			return []string{}
		}
		curr = curr.children[char]
	}
	words := make([]string, 0)
	t.searchPrefixDfs(curr, prefix, &words)
	return words
}
func (t *Trie) searchPrefixDfs(node *TrieNode, path string, words *[]string) {
	if node == nil {
		return
	}

	if node.isEndOfWord {
		*words = append(*words, path)
	}

	for char, child := range node.children {
		t.searchPrefixDfs(child, path+string(char), words)
	}
}

func (t *Trie) StartsWith(prefix string) bool {
	if len(prefix) == 0 {
		return true
	}
	curr := t.root
	for _, char := range prefix {
		if curr.children[char] == nil {
			return false
		}
		curr = curr.children[char]
	}
	return true
}

func (t *Trie) fuzzySearchDfs(node *TrieNode, pattern string, path string, index int, words *[]string) {
	if index == len(pattern) {
		if node.isEndOfWord {
			*words = append(*words, path)
		}
		return
	}

	char := rune(pattern[index])
	switch char {
	case '?':
		for childChar, child := range node.children {
			t.fuzzySearchDfs(child, pattern, path+string(childChar), index+1, words)
		}
	case '*':
		t.fuzzySearchDfs(node, pattern, path, index+1, words)
		for childChar, child := range node.children {
			t.fuzzySearchDfs(child, pattern, path+string(childChar), index, words)
		}
	default:
		if node.children[char] != nil {
			t.fuzzySearchDfs(node.children[char], pattern, path+string(char), index+1, words)
		}

	}
}

func GenerateTrie(words []string) *Trie {
	trie := NewTrie()
	trie.Populate(words)
	return trie
}

File: internal/trie/trie.py


class TrieNode:
    def __init__(self, value=None):
        self.value = value
        self.children = {}
        self.is_end_of_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        if not word:
            return
        curr = self.root
        for ch in word:
            if ch not in curr.children:
                curr.children[ch] = TrieNode(ch)
            curr = curr.children[ch]
        curr.is_end_of_word = True

    def delete(self, word: str):

        def _delete(node: TrieNode, word: str, index: int):
            if index == len(word):
                if not node.is_end_of_word:
                    return False
                node.is_end_of_word = False
                return len(node.children) == 0

            ch = word[index]
            if ch not in node.children:
                return False

            should_delete_child = _delete(node.children[ch], word, index+1)
            if should_delete_child:
                del node.children[ch]
                return len(node.children) == 0
            return False

        return _delete(self.root, word, 0)

    def getAllWords(self):
        if not self.root.children:
            return []
        words = []

        def dfs(node: TrieNode, path: str):
            if node.is_end_of_word:
                words.append(path)

            for ch, child in node.children.items():
                dfs(child, path+ch)

        dfs(self.root, "")
        return words

    def getLongestCommonPrefix(self) -> str:
        prefix = []
        curr = self.root
        while len(curr.children) == 1 and not curr.is_end_of_word:
            ch = next(iter(curr.children))
            prefix.append(ch)
            curr = curr.children[ch]
        return "".join(prefix)

    def fuzzySearch(self, pattern: str):
        if not pattern:
            return []

        words = []

        def dfs(node: TrieNode, path: str, index: int):
            if index == len(pattern):
                if node.is_end_of_word:
                    words.append(path)
                return

            ch = pattern[index]
            if ch == "?":
                for ch, child in node.children.items():
                    dfs(child, path+ch, index+1)
            elif ch == "*":
                dfs(node, path, index + 1)
                for child_ch, child in node.children.items():
                    dfs(child, path + child_ch, index)
            else:
                if ch in node.children:
                    dfs(node.children[ch], path + ch, index + 1)

        dfs(self.root, "", 0)
        return words

    def populate(self, words: list[str]):
        if not words:
            return
        for word in words:
            self.insert(word)

    def startsWith(self, prefix: str) -> bool:
        if not prefix:
            return True

        curr = self.root
        for ch in prefix:
            if ch not in curr.children:
                return False
            curr = curr.children[ch]
        return True

    def searchPrefix(self, prefix: str) -> list[str]:
        if not prefix:
            return []
        curr = self.root
        for ch in prefix:
            if ch not in curr.children:
                return []
            curr = curr.children[ch]

        words = []

        def dfs(node: TrieNode, path: str):
            if not node:
                return

            if node.is_end_of_word:
                words.append(path)

            for ch, child in node.children.items():
                dfs(child, path+ch)

        dfs(curr, prefix)
        return words
